applications/FluidDynamicsApplication/custom_conditions/adjoint_monolithic_wall_condition.cpp

```diff
@@ -0,0 +1,779 @@
+                         std::pow(u_tau / velocity_norm, 2) * velocity_norm_derivative);
+
+                    for (IndexType i = 0; i < TDim; ++i) {
+                        rLocalMatrix(derivative_row, c * TBlockSize + i) -=
+                            velocity[i] * coeff_derivative;
+                    }
+
+                    rLocalMatrix(derivative_row, derivative_row) -= coeff;
+                }
+            }
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::ApplyWallLawShapeDerivatives(
+    MatrixType& rLocalMatrix,
+    const ProcessInfo& rCurrentProcessInfo) const
+{
+    KRATOS_TRY
+
+    const auto& r_geometry = this->GetGeometry();
+
+    const auto& aux_normal = this->GetValue(NORMAL);
+    array_1d<double, TDim> normal;
+    for (IndexType i = 0; i < TDim; ++i) {
+        normal[i] = aux_normal[i];
+    }
+    const double area = norm_2(normal) / TNumNodes;
+    const auto& normal_derivatives = r_geometry.GetValue(NORMAL_SHAPE_DERIVATIVE);
+    const BoundedVector<double, TCoordsLocalSize>& area_derivatives =
+        prod(normal_derivatives, normal) / (area * TNumNodes * TNumNodes);
+
+    const double kappa = 0.41;
+    const double limit_y_plus = 10.9931899;
+    const double beta = 5.2;
+
+    for (IndexType i = 0; i < TNumNodes; ++i) {
+        const auto& r_node = r_geometry[i];
+        const double y = r_node.GetValue(Y_WALL);
+
+        if (y > 0.0 && r_node.Is(SLIP)) {
+            array_1d<double, 3> velocity = r_node.FastGetSolutionStepValue(VELOCITY);
+            velocity -= r_node.FastGetSolutionStepValue(MESH_VELOCITY);
+            const double nu = r_node.FastGetSolutionStepValue(VISCOSITY);
+            const double rho = r_node.FastGetSolutionStepValue(DENSITY);
+            const double velocity_norm = norm_2(velocity);
+
+            if (velocity_norm > 1e-12) {
+                const double y_plus = FluidCalculationUtilities::CalculateLogarithmicYPlus(
+                    velocity_norm, y, nu, kappa, beta, limit_y_plus, 100);
+                const double u_tau = y_plus * nu / y;
+
+                IndexType local_index = 0;
+                for (IndexType c = 0; c < TNumNodes; ++c) {
+                    for (IndexType k = 0; k < TDim; ++k) {
+                        const double coeff_derivative =
+                            area_derivatives[local_index] * u_tau * u_tau * rho / velocity_norm;
+                        for (IndexType j = 0; j < TDim; ++j) {
+                            rLocalMatrix(local_index, i * TBlockSize + j) -=
+                                velocity[j] * coeff_derivative;
+                        }
+                        ++local_index;
+                    }
+                }
+            }
+        }
+    }
+
+    KRATOS_CATCH("");
+}
+
+template class AdjointMonolithicWallCondition<2, 2>;
+template class AdjointMonolithicWallCondition<3, 3>;
+
```

Comment:
Just mentioning that at some point we should start using another flag for the walls as `SLIP` could be misleading (with this what comes to my mind is a pure slip condition rather than a wall law one).

This is of course to be discussed out of this PR.

Reply:
Yes I totally agree. Currently I have `RANS_IS_WALL_FUNCTION_ACTIVE` in RANS which I would like to get rid of eventually :)